lib/internal/http.js

```diff
@@ -121,59 +162,93 @@ class ProxyConfig {
-        if (host.endsWith(suffix)) return false;
+  #buildBypassMatchFns(bypassList = this.#bypassList) {
+    this.#bypassMatchFns = [];
+
+    for (const entry of this.#bypassList) {
+      if (
+        // Handle wildcard entries like *.example.com
+        entry.startsWith('*.') ||
+        // Follow curl's behavior: strip leading dot before matching suffixes.
+        entry.startsWith('.')
+      ) {
+        const suffix = entry.split('');
+        suffix.shift(); // Remove the leading dot or asterisk.
+        const suffixLength = suffix.length;
+        if (suffixLength === 0) {
+          // If the suffix is empty, it means to match all hosts.
+          this.#bypassMatchFns.push(() => true);
+          continue;
+        }
+        this.#bypassMatchFns.push((host) => {
+          const offset = host.length - suffixLength;
+          if (offset < 0) return false; // Host is shorter than the suffix.
+          for (let i = 0; i < suffixLength; i++) {
+            if (host[offset + i] !== suffix[i]) {
+              return false;
+            }
+          }
+          return true;
+        });
+        continue;
       }
 
-      // Handle wildcards like *.example.com
-      if (entry.startsWith('*.') && host.endsWith(entry.substring(1))) return false;
-
       // Handle IP ranges (simple format like 192.168.1.0-192.168.1.255)
       // TODO(joyeecheung): support IPv6.
-      if (entry.includes('-') && isIPv4(host)) {
-        let { 0: startIP, 1: endIP } = entry.split('-');
-        startIP = startIP.trim();
-        endIP = endIP.trim();
-        if (startIP && endIP && isIPv4(startIP) && isIPv4(endIP)) {
-          const hostInt = ipToInt(host);
-          const startInt = ipToInt(startIP);
-          const endInt = ipToInt(endIP);
-          if (hostInt >= startInt && hostInt <= endInt) return false;
-        }
+      const { 0: startIP, 1: endIP } = entry.split('-').map((ip) => ip.trim());
+      if (entry.includes('-') && startIP && endIP && isIPv4(startIP) && isIPv4(endIP)) {
+        const startInt = ipToInt(startIP);
+        const endInt = ipToInt(endIP);
+        this.#bypassMatchFns.push((host) => {
+          if (isIPv4(host)) {
+            const hostInt = ipToInt(host);
+            return hostInt >= startInt && hostInt <= endInt;
+          }
+          return false;
+        });
+        continue;
       }
 
-      // It might be useful to support CIDR notation, but it's not so widely supported
-      // in other tools as a de-facto standard to follow, so we don't implement it for now.
+      // Handle simple host or IP entries
+      this.#bypassMatchFns.push((host, hostWithPort) => {
+        return (host === entry || hostWithPort === entry);
+      });
+    }
+  }
+
+  get bypassList() {
+    // Return a copy of the bypass list to prevent external modification.
+    return [...this.#bypassList];
+  }
+
+  // See: https://about.gitlab.com/blog/we-need-to-talk-no-proxy
+  // TODO(joyeecheung): share code with undici.
```

Comment:
I would prefer this to happen before we do any optimisation. IMO it's better to put this in undici and we can vendor it in in Node.js. It's not ideal if fetch proxy decision starts to diverge from http/https built-ins as we keep making them more convoluted.